mtags/src/main/scala-3/scala/meta/internal/pc/MetalsDriver.scala

```diff
@@ -0,0 +1,42 @@
+package scala.meta.internal.pc
+
+import java.net.URI
+
+import scala.collection.concurrent.TrieMap
+
+import scala.meta.internal.mtags.MD5
+
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.interactive.InteractiveDriver
+import dotty.tools.dotc.interactive.SourceTree
+import dotty.tools.dotc.reporting.Diagnostic
+import dotty.tools.dotc.util.SourceFile
+
+class MetalsDriver(
+    override val settings: List[String]
+) extends InteractiveDriver(settings):
+
+  private val lastCompiled: TrieMap[URI, String] = TrieMap.empty
```

Comment:
I'm not sure which is better
- Storing content as `Array[Char]` and compare them by `Array.equals(...)` vs
- Hash content by md5 (or something) and compare
  - md5 consumes much less memory, but it requires encryption each time
  - but `lastCompiled` will be short-lived ones (as it will be expired when users modify a file), and memory footprint may not matter.

Reply:
Could we rely on didChange/didOpen notifications and force recompilation if the method gets invoked? We wouldn't need to compute anything in that case. so should be faster.

Reply:
> Could we rely on didChange/didOpen notifications and force recompilation if the method gets invoked?

Yeah, I think we can do that!
However, I prefer not to rely on those `didChange/didOpen` events because

- hash comparison (or `Array.equals`) is already pretty fast
  - I commented about `MD5` vs `Array.equals`, but it's about micro-optimization. TBH, either is practically fine (I was just curious) 😅 
- Cache invalidation by those events makes the source code more complicated, and even harder to troubleshoot.

Reply:
ok sure, I was just worried that maybe the hash comparison was problematic :sweat_smile:

Reply:
@tanishiking you could use `TimerProvider` to measure how much time computing hash takes (but maybe log in on debug channel?)

Reply:
I think it would be better at least to attach `didClose` notifications to cleanup cache.
Hypothetically, if you have a project in broken state for a long time it might took some memory ) all units will be in memory until compiler reload, which happens only if bsp compilation pass

Reply:
End up comparing content `Array[Char]` by `ju.Arrays.equals` as Scala2 does. Now `MetalsDriver` compare with the previous one by

```scala
  private def alreadyCompiled(uri: URI, content: Array[Char]): Boolean =
    compilationUnits.get(uri) match
      case Some(unit) if ju.Arrays.equals(unit.source.content(), content) =>
        true
      case _ => false
```

I think this is better because, now we can force re-compile if the driver somehow deleted the compilation unit. (The previous implementation may break the consistency between cache (in `MetalsDriver` and `compilationUnits`).

---

@dos65 
> happens only if bsp compilation pass

I don't think so, presentation compiler instance will be recreated even if BSP compilation fails https://github.com/scalameta/metals/blob/3ca41c31074b91fc4b0c7af0f98a593fa626cd74/metals/src/main/scala/scala/meta/internal/metals/Compilers.scala#L238-L240